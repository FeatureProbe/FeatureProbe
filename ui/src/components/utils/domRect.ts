export class DOMRect {
    x: number;
    width: number;
    y: number;
    height: number;
    constructor(x: number, y: number, width: number, height: number) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }
}