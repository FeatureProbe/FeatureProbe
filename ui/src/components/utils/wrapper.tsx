import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
