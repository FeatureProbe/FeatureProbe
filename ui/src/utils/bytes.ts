export function getBytes(str: string) {
  let bytes = str.length,
    i = 0;
  for (; i < bytes; i++) if (str.charCodeAt(i) > 255) bytes++;

  return bytes;
}

const MB = 1048576;
export const MAX_SIZE = MB;