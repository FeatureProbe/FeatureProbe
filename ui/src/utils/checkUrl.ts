function escapeStringRegexp(string: string): string {
  if (typeof string !== 'string') {
    throw new TypeError('Expected a string');
  }

  return string
    .replace(/[|\\{}()[\]^$+*?.]/g, '\\$&')
    .replace(/-/g, '\\x2d');
}

export function matchUrl(matcher: string, originUrl: string, targetUrl: string): boolean {
  let href = '';
  let hash = '';
  let search = '';

  try {
    const fullUrl = new URL(targetUrl, '');
    href = fullUrl.href;
    hash = fullUrl.hash;
    search = fullUrl.search;
  } catch (error) {
    href = targetUrl;
  }

  let regex;
  let testUrl;

  switch (matcher) {
    case 'EXACT':
      testUrl = href;
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SIMPLE':
      testUrl = href.replace(hash, '').replace(search, '');
      regex = new RegExp('^' + escapeStringRegexp(originUrl) + '/?$');
      break;
    case 'SUBSTRING':
      testUrl = href.replace(search, '');
      regex = new RegExp('.*' + escapeStringRegexp(originUrl) + '.*$');
      break;
    case 'REGULAR':
      testUrl = href.replace(search, '');
      regex = new RegExp(originUrl);
      break;
    default:
      return false;
  }

  return regex.test(testUrl);
}