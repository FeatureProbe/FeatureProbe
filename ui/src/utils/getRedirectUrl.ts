import { getProjectList } from 'services/project';
import { IProject, IEnvironment } from 'interfaces/project';
import { IDictionary } from 'interfaces/targeting';
import { getFromDictionary } from 'services/dictionary';
import { LAST_SEEN } from 'constants/dictionary_keys';

const getInfo = async () => {
  const res =  await getFromDictionary<IDictionary>(LAST_SEEN);
  const { success, data } = res;
  if (success && data) {
    return JSON.parse(data.value);
  } else {
    return {};
  }
};

export const getRedirectUrl = async (defaultRedirectUrl: string) => {
  let redirectUrl = defaultRedirectUrl;
  const info = await getInfo();
  const { projectKey, environmentKey } = info;

  if (!!projectKey && !!environmentKey) {
    redirectUrl = `/${projectKey}/${environmentKey}/toggles`;
  } else {
    const res = await getProjectList<IProject[]>();
    const { success, data } = res;
    if (success && data) {
      if (data[0]) {
        const { key, environments } = data[0];
        let envKey = '';
        environments.some((environment: IEnvironment) => {
          if (environment.key) {
            envKey = environment.key;
            return true;
          }
          return environment;
        });
  
        if (key && envKey) {
          redirectUrl = `/${key}/${envKey}/toggles`;
        }
      } else {
        redirectUrl = '/projects';
      }
    }
  }

  return redirectUrl;
};