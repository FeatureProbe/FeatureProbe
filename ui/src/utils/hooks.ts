
import { useState } from 'react';

export const useLocalStorage = (key: string, initialValue: boolean | string) => {
  const [storedValue, setStoredValue] = useState(() => {
    if (typeof window === 'undefined') {
      return initialValue;
    }

    try {
      const item = window.localStorage.getItem(key);
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      console.log(error);
      return initialValue;
    }
  });

  const setValue = (value: object) => {
    try {
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);

      if (typeof window !== 'undefined') {
        window.localStorage.setItem(key, JSON.stringify(valueToStore));
      }
      
    } catch (error) {
      console.log(error);
    }
  };
  return [storedValue, setValue];
};
