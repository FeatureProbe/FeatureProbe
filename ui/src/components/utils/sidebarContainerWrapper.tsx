import { SidebarContainer } from 'layout/hooks';

export default SidebarContainer.Provider;
