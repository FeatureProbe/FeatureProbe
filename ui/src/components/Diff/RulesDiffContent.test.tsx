import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { RulesDiffContent } from './RulesDiffContent';
import { IntlWrapper } from 'components/utils/wrapper';
import idiff from './diff';

const content_a = idiff(
  [
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is not any of',
          objects: ['aaa', '222'],
        },
        {
          type: 'number',
          subject: 'num',
          predicate: '=',
          objects: ['222'],
        },
        {
          type: 'datetime',
          subject: 'time',
          predicate: 'before',
          objects: ['2023-01-10T10:14:06+08:00'],
        },
        {
          type: 'semver',
          subject: 'ver',
          predicate: '!=',
          objects: ['0.0.1'],
        },
        {
          type: 'segment',
          predicate: 'is not in',
          objects: ['_cn'],
        },
      ],
      name: 'Rule 1',
      serve: {
        select: 0,
      },
    },
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is one of',
          objects: ['ioio'],
        },
      ],
      name: 'Rule2',
      serve: {
        split: [10000, 0],
      },
    },
  ],
  [
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is not any of',
          objects: ['222'],
        },
        {
          type: 'number',
          subject: 'num',
          predicate: '=',
          objects: ['222', '333'],
        },
        {
          type: 'semver',
          subject: 'ver',
          predicate: '!=',
          objects: ['0.0.1'],
        },
        {
          type: 'segment',
          predicate: 'is not in',
          objects: ['_cn'],
        },
        {
          type: 'string',
          subject: 'str2',
          predicate: 'is not any of',
          objects: ['333'],
        },
      ],
      name: 'Rule 1M',
      serve: {
        split: [5000, 5000],
      },
    },
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is one of',
          objects: ['ioio'],
        },
      ],
      name: 'Rule2',
      serve: {
        select: 0,
      },
    },
  ]
);

const content_b = idiff(
  [
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is one of',
          objects: ['ioio'],
        },
      ],
      name: 'Rule2',
      serve: {
        split: [10000, 0],
      },
    },
  ],
  [
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is one of',
          objects: ['ioio'],
        },
      ],
      name: 'Rule2',
      serve: {
        split: [10000, 0],
      },
    },
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is one of',
          objects: ['ioio'],
        },
      ],
      name: 'Rule3',
      serve: {
        split: [10000, 0],
      },
    },
  ]
);

const content_c = idiff(
  [
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is one of',
          objects: ['ioio'],
        },
      ],
      name: 'Rule2',
      serve: {
        split: [10000, 0],
      },
    },
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is one of',
          objects: ['ioio'],
        },
      ],
      name: 'Rule3',
      serve: {
        split: [10000, 0],
      },
    },
  ],
  [
    {
      conditions: [
        {
          type: 'string',
          subject: 'str',
          predicate: 'is one of',
          objects: ['ioio'],
        },
      ],
      name: 'Rule2',
      serve: {
        split: [10000, 0],
      },
    },
  ]
);

it('RulesDiffContent snapshot', (done) => {
  (async () => {
    const { asFragment, rerender } = render(<RulesDiffContent content={content_a} />, {
      wrapper: IntlWrapper,
    });
    expect(asFragment()).toMatchSnapshot();

    rerender(<RulesDiffContent content={content_b} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(<RulesDiffContent content={content_c} />);
    expect(asFragment()).toMatchSnapshot();


    done();
  })();
});
