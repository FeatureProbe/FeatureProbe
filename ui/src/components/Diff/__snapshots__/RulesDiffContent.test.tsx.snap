// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RulesDiffContent snapshot 1`] = `
<DocumentFragment>
  <div
    class="rule-content"
  >
    <div>
      <div>
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class=""
                colspan="6"
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  Rule 1
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                if
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  string
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  str
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  is not any of
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  aaa
                </div>
                <div
                  class="diff-field-value field-value-same"
                  style="background: rgba(33, 37, 41, 0.08);"
                >
                  222
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                and
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  number
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  num
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  =
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                  style="background: rgba(33, 37, 41, 0.08);"
                >
                  222
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-remove condition-diff-item"
            >
              <td
                class=""
              >
                and
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  datetime
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  time
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  before
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-remove"
                  >
                    2023-01-10T10:14:06+08:00
                  </div>
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                and
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  semver
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  ver
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  !=
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-same"
                    style="background: rgba(33, 37, 41, 0.08);"
                  >
                    0.0.1
                  </div>
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                and
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  segment
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  is not in
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-same"
                    style="background: rgba(33, 37, 41, 0.08);"
                  >
                    _cn
                  </div>
                </div>
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              />
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-modify condition-diff-item"
            >
              <td
                class=""
              >
                Serve
              </td>
              <td
                class=""
                colspan="5"
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  0
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div>
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class=""
                colspan="6"
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  Rule2
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                if
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  string
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  str
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  is one of
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-same"
                    style="background: rgba(33, 37, 41, 0.08);"
                  >
                    ioio
                  </div>
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-same condition-diff-item"
            >
              <td
                class=""
              >
                Serve
              </td>
              <td
                class=""
                colspan="5"
              >
                <div>
                  <div
                    class="diff-field-value field-value-remove"
                  >
                    10000
                  </div>
                  <div
                    class="diff-field-value field-value-remove"
                  >
                    0
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div>
      <div>
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class=""
                colspan="6"
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  Rule 1M
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                if
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  string
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  str
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  is not any of
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                  style="background: rgba(33, 37, 41, 0.08);"
                >
                  222
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                and
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  number
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  num
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  =
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                  style="background: rgba(33, 37, 41, 0.08);"
                >
                  222
                </div>
                <div
                  class="diff-field-value field-value-add"
                >
                  333
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              />
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                and
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  semver
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  ver
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  !=
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-same"
                    style="background: rgba(33, 37, 41, 0.08);"
                  >
                    0.0.1
                  </div>
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                and
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  segment
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  is not in
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-same"
                    style="background: rgba(33, 37, 41, 0.08);"
                  >
                    _cn
                  </div>
                </div>
              </td>
              <td
                class=""
              />
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-add condition-diff-item"
            >
              <td
                class=""
              >
                and
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  string
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  str2
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  is not any of
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-add"
                  >
                    333
                  </div>
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-modify condition-diff-item"
            >
              <td
                class=""
              >
                Serve
              </td>
              <td
                class=""
                colspan="5"
              >
                <div>
                  <div
                    class="diff-field-value field-value-add"
                  >
                    5000
                  </div>
                  <div
                    class="diff-field-value field-value-add"
                  >
                    5000
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div>
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class=""
                colspan="6"
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  Rule2
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              >
                if
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  string
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  str
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                >
                  is one of
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-same"
                    style="background: rgba(33, 37, 41, 0.08);"
                  >
                    ioio
                  </div>
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-same condition-diff-item"
            >
              <td
                class=""
              >
                Serve
              </td>
              <td
                class=""
                colspan="5"
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  0
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RulesDiffContent snapshot 2`] = `
<DocumentFragment>
  <div
    class="rule-content"
  >
    <div>
      <div
        class="empty"
      >
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class=""
                colspan="6"
              />
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              />
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-same condition-diff-item"
            >
              <td
                class=""
              >
                Serve
              </td>
              <td
                class=""
                colspan="5"
              >
                <div>
                  <div
                    class="diff-field-value field-value-undefined"
                  >
                    10000
                  </div>
                  <div
                    class="diff-field-value field-value-undefined"
                  >
                    0
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div>
      <div>
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header diff-item-add"
          >
            <tr
              class=""
            >
              <th
                class=""
                colspan="6"
              >
                Rule3
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item-add condition-diff-item"
            >
              <td
                class=""
              >
                if
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  string
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  str
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  is one of
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-add"
                  >
                    ioio
                  </div>
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-add condition-diff-item"
            >
              <td
                class=""
              >
                Serve
              </td>
              <td
                class=""
                colspan="5"
              >
                <div>
                  <div
                    class="diff-field-value field-value-undefined"
                  >
                    10000
                  </div>
                  <div
                    class="diff-field-value field-value-undefined"
                  >
                    0
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RulesDiffContent snapshot 3`] = `
<DocumentFragment>
  <div
    class="rule-content"
  >
    <div>
      <div>
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header diff-item-remove"
          >
            <tr
              class=""
            >
              <th
                class=""
                colspan="6"
              >
                Rule3
              </th>
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item-remove condition-diff-item"
            >
              <td
                class=""
              >
                if
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  string
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  str
                </div>
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  is one of
                </div>
              </td>
              <td
                class=""
              >
                <div>
                  <div
                    class="diff-field-value field-value-remove"
                  >
                    ioio
                  </div>
                </div>
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-remove condition-diff-item"
            >
              <td
                class=""
              >
                Serve
              </td>
              <td
                class=""
                colspan="5"
              >
                <div>
                  <div
                    class="diff-field-value field-value-undefined"
                  >
                    10000
                  </div>
                  <div
                    class="diff-field-value field-value-undefined"
                  >
                    0
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div>
      <div
        class="empty"
      >
        <table
          class="ui small unstackable very basic table"
        >
          <thead
            class="table-header"
          >
            <tr
              class=""
            >
              <th
                class=""
                colspan="6"
              />
            </tr>
          </thead>
          <tbody
            class=""
          >
            <tr
              class="diff-item-normal condition-diff-item"
            >
              <td
                class=""
              />
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              >
                <div
                  class="diff-field-value field-value-same"
                />
              </td>
              <td
                class=""
              />
            </tr>
            <tr
              class="diff-item-same condition-diff-item"
            >
              <td
                class=""
              >
                Serve
              </td>
              <td
                class=""
                colspan="5"
              >
                <div>
                  <div
                    class="diff-field-value field-value-undefined"
                  >
                    10000
                  </div>
                  <div
                    class="diff-field-value field-value-undefined"
                  >
                    0
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
